libcore/UndoManager.vala

```diff
@@ -0,0 +1,919 @@
+
+    public UndoManagerData (Marlin.UndoActionType type, int items_count) {
+        this.type = type;
+        this.count = items_count;
+
+        if (type == Marlin.UndoActionType.MOVETOTRASH) {
+            trashed = new GLib.HashTable<string, uint64?> (str_hash, str_equal);
+        }
+
+        sources = new GLib.List<string> ();
+        destinations = new GLib.List<string> ();
+    }
+
+    public void set_src_dir (GLib.File src) {
+        src_dir = src;
+    }
+
+    public void set_dest_dir (GLib.File dest) {
+        dest_dir = dest;
+    }
+
+    public void add_origin_target_pair (GLib.File origin, GLib.File target) {
+        var src_relative = src_dir.get_relative_path (origin);
+        sources.prepend (src_relative);
+        var dest_relative = dest_dir.get_relative_path (target);
+        destinations.prepend (dest_relative);
+
+        is_valid = true;
+    }
+
+    public void add_trashed_file (GLib.File file, uint64 mtime) {
+        trashed.insert (file.get_uri (), mtime);
+
+        is_valid = true;
+    }
+
+    public void set_create_data (string target_uri, string template) {
+        this.template = template;
+        this.target_uri = target_uri;
+
+        is_valid = true;
+    }
+
+    public void set_rename_information (GLib.File old_file, GLib.File new_file) {
+        old_uri = old_file.get_uri ();
+        new_uri = new_file.get_uri ();
+
+        is_valid = true;
+    }
+
+    public void add_file_permissions (GLib.File file, uint32 permission) {
+        original_permissions.insert (file.get_uri (), permission);
+
+        is_valid = true;
+    }
+
+    public void set_recursive_permissions (uint32 file_permissions, uint32 file_mask, uint32 dir_permissions, uint32 dir_mask) {
+        this.file_permissions = file_permissions;
+        this.file_mask = file_mask;
+        this.dir_permissions = dir_permissions;
+        this.dir_mask = dir_mask;
+
+        is_valid = true;
+    }
+
+    public void set_file_permissions (string uri, uint32 current_permissions, uint32 new_permissions) {
+        target_uri = uri;
+
+        this.current_permissions = current_permissions;
+        this.new_permissions = new_permissions;
+
+        is_valid = true;
+    }
+
+    public void set_owner_change_information (string uri, string current_user, string new_user) {
+        target_uri = uri;
+
+        original_user_name_or_id = current_user;
+        new_user_name_or_id = new_user;
+
+        is_valid = true;
+    }
+
+    public void set_group_change_information (string uri, string current_group, string new_group) {
+        target_uri = uri;
+
+        original_group_name_or_id = current_group;
+        new_group_name_or_id = new_group;
+
+        is_valid = true;
+    }
+
+    private unowned string get_first_target_short_name () {
+        return destinations.first ().data;
+    }
+
+    private string get_uri_parent_path (string uri) {
+        return GLib.File.new_for_uri (uri).get_parent ().get_path ();
```

Comment:
.get_parent () could theoretically return null.

Reply:
Indeed, it wasn't taken into account in the original code but I added a workaround for this issue